File: ALittle/Module/ALittleIDE/Other/GameLibrary/CEngine/Script/CEngine/UISystem/Base/Matrix2D.lua
-- ALittle Generate Lua
module("ALittle", package.seeall)

local ___thispath = select('1', ...):match("(.+[/\\]).+$") or ""
local ___rawset = rawset
local ___pairs = pairs
local ___ipairs = ipairs


local __cos = Math_Cos
local __sin = Math_Sin
Matrix2D = Lua.Class(nil, "ALittle.Matrix2D")

function Matrix2D:Ctor()
	local m = {}
	m[0] = {}
	m[1] = {}
	m[2] = {}
	___rawset(self, "_m", m)
	self:SetIdentity()
end

function Matrix2D:SetIdentity()
	local m = self._m
	m[0][0] = 1
	m[0][1] = 0
	m[0][2] = 0
	m[1][0] = 0
	m[1][1] = 1
	m[1][2] = 0
	m[2][0] = 0
	m[2][1] = 0
	m[2][2] = 1
end

function Matrix2D:SetRotate(rad)
	local m = self._m
	m[0][0] = __cos(rad)
	m[0][1] = __sin(rad)
	m[0][2] = 0
	m[1][0] = -m[0][1]
	m[1][1] = m[0][0]
	m[1][2] = 0
	m[2][0] = 0
	m[2][1] = 0
	m[2][2] = 1
end

function Matrix2D:Rotate(rad)
	local m = Matrix2D()
	m:SetRotate(rad)
	self:Multiply(m)
end

function Matrix2D:SetTranslation(x, y)
	local m = self._m
	m[0][0] = 1
	m[0][1] = 0
	m[0][2] = 0
	m[1][0] = 0
	m[1][1] = 1
	m[1][2] = 0
	m[2][0] = x
	m[2][1] = y
	m[2][2] = 1
end

function Matrix2D:Translation(x, y)
	local m = Matrix2D()
	m:SetTranslation(x, y)
	self:Multiply(m)
end

function Matrix2D:SetScale(x, y)
	local m = self._m
	m[0][0] = x
	m[0][1] = 0
	m[0][2] = 0
	m[1][0] = 0
	m[1][1] = y
	m[1][2] = 0
	m[2][0] = 0
	m[2][1] = 0
	m[2][2] = 1
end

function Matrix2D:Scale(x, y)
	local m = Matrix2D()
	m:SetScale(x, y)
	self:Multiply(m)
end

function Matrix2D:Multiply(right)
	local r = {}
	local m = self._m
	local rm = right._m
	r[0] = {}
	r[0][0] = m[0][0] * rm[0][0] + m[0][1] * rm[1][0] + m[0][2] * rm[2][0]
	r[0][1] = m[0][0] * rm[0][1] + m[0][1] * rm[1][1] + m[0][2] * rm[2][1]
	r[0][2] = m[0][0] * rm[0][2] + m[0][1] * rm[1][2] + m[0][2] * rm[2][2]
	r[1] = {}
	r[1][0] = m[1][0] * rm[0][0] + m[1][1] * rm[1][0] + m[1][2] * rm[2][0]
	r[1][1] = m[1][0] * rm[0][1] + m[1][1] * rm[1][1] + m[1][2] * rm[2][1]
	r[1][2] = m[1][0] * rm[0][2] + m[1][1] * rm[1][2] + m[1][2] * rm[2][2]
	r[2] = {}
	r[2][0] = m[2][0] * rm[0][0] + m[2][1] * rm[1][0] + m[2][2] * rm[2][0]
	r[2][1] = m[2][0] * rm[0][1] + m[2][1] * rm[1][1] + m[2][2] * rm[2][1]
	r[2][2] = m[2][0] * rm[0][2] + m[2][1] * rm[1][2] + m[2][2] * rm[2][2]
	self._m = r
end

Vector2D = Lua.Class(nil, "ALittle.Vector2D")

function Vector2D:Ctor(xx, yy)
	___rawset(self, "_x", xx)
	___rawset(self, "_y", yy)
end

function Vector2D:Multiply(right)
	local rm = right._m
	local xx = self._x * rm[0][0] + self._y * rm[1][0] + 1 * rm[2][0]
	local yy = self._x * rm[0][1] + self._y * rm[1][1] + 1 * rm[2][1]
	self._x = xx
	self._y = yy
end

File: ALittleIDE/Module/ALittleIDE/Other/GameLibrary/CEngine/JSScript/CEngine/UISystem/Base/Matrix2D.js
{
if (typeof ALittle === "undefined") window.ALittle = {};


let __cos = ALittle.Math_Cos;
let __sin = ALittle.Math_Sin;
ALittle.Matrix2D = JavaScript.Class(undefined, {
	Ctor : function() {
		let m = new Map();
		m.set(0, new Map());
		m.set(1, new Map());
		m.set(2, new Map());
		this._m = m;
		this.SetIdentity();
	},
	SetIdentity : function() {
		let m = this._m;
		m.get(0).set(0, 1);
		m.get(0).set(1, 0);
		m.get(0).set(2, 0);
		m.get(1).set(0, 0);
		m.get(1).set(1, 1);
		m.get(1).set(2, 0);
		m.get(2).set(0, 0);
		m.get(2).set(1, 0);
		m.get(2).set(2, 1);
	},
	SetRotate : function(rad) {
		let m = this._m;
		m.get(0).set(0, __cos(rad));
		m.get(0).set(1, __sin(rad));
		m.get(0).set(2, 0);
		m.get(1).set(0, -m.get(0).get(1));
		m.get(1).set(1, m.get(0).get(0));
		m.get(1).set(2, 0);
		m.get(2).set(0, 0);
		m.get(2).set(1, 0);
		m.get(2).set(2, 1);
	},
	Rotate : function(rad) {
		let m = ALittle.NewObject(ALittle.Matrix2D);
		m.SetRotate(rad);
		this.Multiply(m);
	},
	SetTranslation : function(x, y) {
		let m = this._m;
		m.get(0).set(0, 1);
		m.get(0).set(1, 0);
		m.get(0).set(2, 0);
		m.get(1).set(0, 0);
		m.get(1).set(1, 1);
		m.get(1).set(2, 0);
		m.get(2).set(0, x);
		m.get(2).set(1, y);
		m.get(2).set(2, 1);
	},
	Translation : function(x, y) {
		let m = ALittle.NewObject(ALittle.Matrix2D);
		m.SetTranslation(x, y);
		this.Multiply(m);
	},
	SetScale : function(x, y) {
		let m = this._m;
		m.get(0).set(0, x);
		m.get(0).set(1, 0);
		m.get(0).set(2, 0);
		m.get(1).set(0, 0);
		m.get(1).set(1, y);
		m.get(1).set(2, 0);
		m.get(2).set(0, 0);
		m.get(2).set(1, 0);
		m.get(2).set(2, 1);
	},
	Scale : function(x, y) {
		let m = ALittle.NewObject(ALittle.Matrix2D);
		m.SetScale(x, y);
		this.Multiply(m);
	},
	Multiply : function(right) {
		let r = new Map();
		let m = this._m;
		let rm = right._m;
		r.set(0, new Map());
		r.get(0).set(0, m.get(0).get(0) * rm.get(0).get(0) + m.get(0).get(1) * rm.get(1).get(0) + m.get(0).get(2) * rm.get(2).get(0));
		r.get(0).set(1, m.get(0).get(0) * rm.get(0).get(1) + m.get(0).get(1) * rm.get(1).get(1) + m.get(0).get(2) * rm.get(2).get(1));
		r.get(0).set(2, m.get(0).get(0) * rm.get(0).get(2) + m.get(0).get(1) * rm.get(1).get(2) + m.get(0).get(2) * rm.get(2).get(2));
		r.set(1, new Map());
		r.get(1).set(0, m.get(1).get(0) * rm.get(0).get(0) + m.get(1).get(1) * rm.get(1).get(0) + m.get(1).get(2) * rm.get(2).get(0));
		r.get(1).set(1, m.get(1).get(0) * rm.get(0).get(1) + m.get(1).get(1) * rm.get(1).get(1) + m.get(1).get(2) * rm.get(2).get(1));
		r.get(1).set(2, m.get(1).get(0) * rm.get(0).get(2) + m.get(1).get(1) * rm.get(1).get(2) + m.get(1).get(2) * rm.get(2).get(2));
		r.set(2, new Map());
		r.get(2).set(0, m.get(2).get(0) * rm.get(0).get(0) + m.get(2).get(1) * rm.get(1).get(0) + m.get(2).get(2) * rm.get(2).get(0));
		r.get(2).set(1, m.get(2).get(0) * rm.get(0).get(1) + m.get(2).get(1) * rm.get(1).get(1) + m.get(2).get(2) * rm.get(2).get(1));
		r.get(2).set(2, m.get(2).get(0) * rm.get(0).get(2) + m.get(2).get(1) * rm.get(1).get(2) + m.get(2).get(2) * rm.get(2).get(2));
		this._m = r;
	},
}, "ALittle.Matrix2D");

ALittle.Vector2D = JavaScript.Class(undefined, {
	Ctor : function(xx, yy) {
		this._x = xx;
		this._y = yy;
	},
	Multiply : function(right) {
		let rm = right._m;
		let xx = this._x * rm.get(0).get(0) + this._y * rm.get(1).get(0) + 1 * rm.get(2).get(0);
		let yy = this._x * rm.get(0).get(1) + this._y * rm.get(1).get(1) + 1 * rm.get(2).get(1);
		this._x = xx;
		this._y = yy;
	},
}, "ALittle.Vector2D");

}